### PR TITLE
WrappedSynchronizer

### LINKED ISSUE #291: Wrapped Synchronzier
The current CombineSynchronzier and way how publishing new values is done can be reimplemented to much simpler and testable version. 

### The issue(s) and solution(s)
**Publishing Issue**
Every time synchronzier updates the method updatePublishers is called. That one gets values for all publishers and broadcast those even though user can be looking at screen that doesn't need any of it. 

**Publishing Solutiuon**
Let's reimplement it so we inform any interested part by broadcasting synchronizerSynced as the only information. Anyone interested needs to subscribe to get know it and unsubscribes once it doesn't it anymore. 
The values needed will be requested as side effects for the requested values only. 

**Naming Issue**
The CombineSynchronizer is not very much Combine one, we don't have any other Synchronizers like RxSwift or similar, the Combine also doesn't reflect exactly the purpose. Technically it's more about wrapping Synchronizer. 

**Naming Solution**
We should then use WrappedSynchronizer or SynchronizerInteractor.

NOTE: when .synchronizerSynced is sent let's define possible stated of the synchronizes and pass it along. It should be all states from SDKSycnhronizer.swift file under public extension Notification.Name {.

### PR DESCRIPTION
**Pull request by LukasKorba**
Closes #291 

This code review checklist is intended to serve as a starting point for the author and reviewer, although it may not be appropriate for all types of changes (e.g. fixing a spelling typo in documentation).  For more in-depth discussion of how we think about code review, please see [Code Review Guidelines](../blob/main/CODE_REVIEW_GUIDELINES.md).

# Author

- [x] Self-review: Did you review your own code in GitHub's web interface? Code often looks different when reviewing the diff in a browser, making it easier to spot potential bugs.
- [x] Does the code abide by the [Coding Guidelines](../blob/main/docs/CODING_GUIDELINES.md)?
- [ ] Automated tests: Did you add appropriate automated tests for any code changes?
- [x] Code coverage: Did you check the code coverage report for the automated tests?  While we are not looking for perfect coverage, the tool can point out potential cases that have been missed.
- [ ] Documentation: Did you update Docs as appropiate? (E.g [README.md](../blob/main/README.md), etc.)
- [x] Run the app: Did you run the app and try the changes? 
- [ ] Did you provide Screenshots of what the App looks like before and after your changes as part of the description of this PR? (only applicable to UI Changes)
- [x] Rebase and squash: Did you pull in the latest changes from the main branch and squash your commits before assigning a reviewer? Having your code up to date and squashed will make it easier for others to review. Use best judgement when squashing commits, as some changes (such as refactoring) might be easier to review as a separate commit.

# Reviewer

- [x] Checklist review: Did you go through the code with the [Code Review Guidelines](../blob/main/CODE_REVIEW_GUIDELINES.md) checklist?
- [x] Ad hoc review: Did you perform an ad hoc review?  _In addition to a first pass using the code review guidelines, do a second pass using your best judgement and experience which may identify additional questions or comments. Research shows that code review is most effective when done in multiple passes, where reviewers look for different things through each pass._
- [ ] Automated tests: Did you review the automated tests?
- [ ] Manual tests: Did you review the manual tests?_You will find manual testing guidelines under our [manual testing section](../blob/main/docs/testing/manual_testing)_
- [ ] How is Code Coverage affected by this PR? _We encourage you to compare coverage before and after your changes and when possible, leave it in a better place. [Learn More...](../blob/main/docs/testing/local_coverage.md)_
- [ ] Documentation: Did you review Docs, [README.md](../blob/main/README.md), [LICENSE.md](../blob/main/LICENSE.md), and [Architecture.md](../blob/main/docs/Architecture.md) as appropriate?
- [ ] Run the app: Did you run the app and try the changes? While the CI server runs the app to look for build failures or crashes, humans running the app are more likely to notice unexpected log messages, UI inconsistencies, or bad output data.

### CODE REVIEW

File: secant/Wrappers/WrappedSynchronizer.swift [hunk 1]
```diff
@@ -8,26 +8,51 @@
...
+    case stopped
+    case disconnected
+    case syncing
+    case downloading
+    case validating
+    case scanning
+    case enhancing
+    case fetching
+    case minedTransaction
+    case foundTransactions
```
Comment (pacu) [inline secant/Wrappers/WrappedSynchronizer.swift:30]: What would be the difference between mined transaction and foundTransactions?

We are currently not using the mempool to fetch unmined transactions so every transaction not originated on this wallet will be detected when mined.

**Review by pacu — CHANGES_REQUESTED** [review]

File: secant/Wrappers/WrappedSynchronizer.swift (hunk 1 above)
Comment (LukasKorba) [reply secant/Wrappers/WrappedSynchronizer.swift:30]: as agreed on slack, we need to support both

File: secant/Wrappers/WrappedSynchronizer.swift [hunk 2]
```diff
@@ -8,26 +8,51 @@
...
+    case connectionStateChanged
+}
 
 struct Balance: WalletBalance, Equatable {
     var verified: Int64
     var total: Int64
 }
 
-protocol CombineSynchronizer {
+protocol WrappedSynchronizer {
```
Comment (pacu) [inline secant/Wrappers/WrappedSynchronizer.swift:40]: I had made this comment but I don't see it anymore. I'm not sure this is the correct name for this protocol. Why is the synchronizer being wrapped? Shouldn't that be reflected in the name of the protocol? 

Wouldn't `ComposableSynchronizer` or something along those lines be more suitable?

**Review by pacu — CHANGES_REQUESTED** [review]
Adding a comment that somehow got lost.

File: secant/Wrappers/WrappedSynchronizer.swift (hunk 2 above)
Comment (LukasKorba) [reply secant/Wrappers/WrappedSynchronizer.swift:40]: The naming is challenging, I know. Technically it's "wrapping" the SDKSynchronizer
```swift
class LiveWrappedSynchronizer: WrappedSynchronizer {
    private var cancellables: [AnyCancellable] = []
    private(set) var synchronizer: SDKSynchronizer?
```

`ComposableSynchronizer` is not fit as well, there is no real composable in it. Neither the TCA, neither the composability. This synchronizer is basically a dependency. Object providing interface to get some data in a form of Effects. It is not much different from WrappedMnemonic, WrappedFileManager, etc. It's using some live object iunder the hood to interact with, hence wrapped.

I'll give it today to think something else but WrappedSynchronizer is I believe "the best" for now.

Comment (LukasKorba) [reply secant/Wrappers/WrappedSynchronizer.swift:40]: After it settled, I think the naming misses the context. What if we add SDK to it? 
```swift
protocol WrappedSDKSynchronizer
```
and then we simply define (exclude wrapped)
```swift
class LiveSDKSynchronizer: WrappedSDKSynchronizer { }

class MockSDKSynchronizer: WrappedSDKSynchronizer { }
```

**Review by pacu — APPROVED** [review]

**Pull request merged**